app/uk/gov/hmrc/gform/upscan/UpscanService.scala

```diff
@@ -0,0 +1,130 @@
+import uk.gov.hmrc.gform.config.AppConfig
+import uk.gov.hmrc.gform.gformbackend.GformConnector
+import uk.gov.hmrc.gform.sharedmodel.AccessCode
+import uk.gov.hmrc.gform.sharedmodel.form.{ EnvelopeId, Form, FormIdData, UserData }
+import uk.gov.hmrc.gform.sharedmodel.formtemplate.{ FormComponentId, FormTemplateId, SectionNumber }
+import uk.gov.hmrc.gform.upscan.routes.UpscanController
+import uk.gov.hmrc.http.HeaderCarrier
+
+class UpscanService(
+  upscanConnector: UpscanConnector,
+  upscanRepository: UpscanRepository,
+  gformConnector: GformConnector,
+  queryParameterCrypto: CryptoWithKeysFromConfig,
+  appConfig: AppConfig
+)(implicit ec: ExecutionContext)
+    extends UpscanAlgebra[Future] {
+
+  def upscanInitiate(
+    fileUploadIds: List[FormComponentId],
+    formTemplateId: FormTemplateId,
+    sectionNumber: SectionNumber,
+    maybeAccessCode: Option[AccessCode],
+    form: Form,
+    formIdData: FormIdData
+  )(implicit
+    hc: HeaderCarrier
+  ): Future[UpscanInitiate] =
+    fileUploadIds
+      .traverse(formComponentId =>
+        upscanConnector
+          .upscanInitiate(
+            toRequest(formTemplateId, sectionNumber, maybeAccessCode, formComponentId, form.envelopeId, formIdData)
+          )
+          .map(formComponentId -> _)
+      )
+      .flatMap { fcIdWithResponse =>
+        val upscanInititate = UpscanInitiate(fcIdWithResponse.toMap)
+
+        val formIdData: FormIdData = FormIdData.fromForm(form, maybeAccessCode)
+        val userData: UserData = UserData(
+          form.formData,
+          form.status,
+          form.visitsIndex,
+          form.thirdPartyData,
+          form.componentIdToFileId
+        )
+
+        gformConnector.updateUserData(formIdData, userData).map(_ => upscanInititate)
+
+      }
+
+  private def toRequest(
+    formTemplateId: FormTemplateId,
+    sectionNumber: SectionNumber,
+    maybeAccessCode: Option[AccessCode],
+    formComponentId: FormComponentId,
+    envelopeId: EnvelopeId,
+    formIdData: FormIdData
+  ): UpscanInitiateRequest = {
+
+    val baseUrl = appConfig.`gform-frontend-base-url`
+
+    val formIdDataString = Json.stringify(Json.toJson(formIdData))
+
+    val formIdDataCrypted: Crypted = queryParameterCrypto.encrypt(PlainText(formIdDataString))
+
+    val callback: String =
+      baseUrl + UpscanController.callback(formComponentId, envelopeId, formIdDataCrypted).url
+    val successRedirect: String =
+      baseUrl + UpscanController.success(formTemplateId, sectionNumber, maybeAccessCode, formComponentId).url
+    val errorRedirect: String =
+      baseUrl + UpscanController.error(formTemplateId, sectionNumber, maybeAccessCode, formComponentId).url
+
+    UpscanInitiateRequest(
+      callback,
+      successRedirect,
+      errorRedirect
+    )
+  }
+
+  def download(
+    downloadUrl: String
+  ): Future[ByteString] = upscanConnector.download(downloadUrl)
+
+  def waitForConfirmation(reference: UpscanReference): Future[UpscanFileStatus] =
```

Comment:
Why call this waitForConfirmation? Is the find method going to be blocked until a value appears in the mongo collection?

Reply:
renamed to `retrieveConfirmationOrFail`